### ISSUE TITLE
Big creatures or items are not rendered when covered by upper floor

### ISSUE THREAD

**Issue by oarsc** [opentibiabr/otclient#120]
### Before creating an issue, please ensure:
- [x] This is a bug in the software that resides in this repository, and not a
      support matter (use https://otland.net/forums/otclient.494/ for support)
- [x] This issue is reproducible without changes to the code in this repository

### Steps to reproduce (include any configuration/script required to reproduce)
1. See video included

### Expected behaviour
Creatures or items should be render partially covered by the floor

### Actual behaviour
https://i.imgur.com/U8m2ZUx.mp4

### Environment
OTHire 7.72
Ubuntu 18.04

**Comment by oarsc** [comment]
Hello, it's already solved when the creature/item is not moving, but 
monsters just pop out/in when they move from a total coverage to a partially one.
Video: https://i.imgur.com/oPeB1xj.mp4

**Comment by oarsc** [comment]
I'm sorry to say that it's still happening sometimes:
https://i.imgur.com/w8oAsHE.mp4

Watch the full video, when you relogin it fixes temporally, unless you move something

**Comment by mehah** [comment]
I just tested it and it's working perfectly, have you applied all the fixes?

look this method: https://github.com/mehah/otclient/blob/68f3c46a43e8c4e94e585ba9418c21933994867f/src/client/tile.cpp#L252

https://user-images.githubusercontent.com/2267386/127574342-cf53fa53-0a1d-4434-8737-0ab36825a18e.mp4

**Comment by oarsc** [comment]
Yes, I have the all commits from "cache-for-all"
It happens in some occasions, where observer point of view changes from being able to see seeing under to not being able:
https://i.imgur.com/PQuT1Kh.mp4 (a bit low quality, but you can see it happening)

This is how I found I can reproduce it with monsters (minute 0:22 and 0:43 of my video):
1- Monster being TOTALLY COVERED from a roof or higher level
2- Player logs in (or comes from away into the scene)
3- Player moves near to see the monster (so he can see what's under that roof)
4- Monster moves to the partially covered spot
5- Player gets away to see the roof
6- The monster is invisible...

Another way to making it happening (minute 0:05)
1- Player logs in (or comes from away into the scene)
2- Monster logs in (if that is even possible xD) in a partially covered spot
3- The monster is invisible

If the player has already seen the monster partially covered, it doesn't matter if the monster "relogins", the player will still see it (minute 0:12)

**Comment by oarsc** [comment]
Ok, I found out that you don't need to login or something:
https://user-images.githubusercontent.com/5507777/127577575-a65dacaf-e2b8-4cd6-a452-9844d7ccaa79.mp4

Steps are:
1- Monster covers totally under roof so player can't see it.
2- Player moves to see under roof
3- Monster moves to the partially covered spot
4- Player moves back to his position

If you swap any step, it'll work fine

**Comment by mehah** [comment]
see if now is fixed :)

**Comment by oarsc** [comment]
Mmmm, the step I mentioned in my last comment seem to work but, looks like there is still something off...
I could see something strange login on a partially covered position:
https://www.youtube.com/watch?v=T-U__B219Z4 watch the full video
(Had to upload to youtube cuz I recorded way too much for github/imgur...)

Looks like it is affected by the position, and the position where it comes from... When moving from the south tile, turning didn't make the demon disapear :S:S

**Comment by mehah** [comment]
what's on top, is it full ground?

**Comment by mehah** [comment]
I'm not able to simulate here, if it's really causing a problem in your environment, I'll end up going back to the way it was before, since this problem doesn't occur here in my test environment.

**Comment by mehah** [comment]
I did everything you did in the video and it didn't happen, are you using the clean version, if not, have you applied all the changes?

take a look at these lines:

https://github.com/mehah/otclient/blob/3c90b17379ad690f64d596c054aa75830a66fe9e/src/client/tile.cpp#L331
https://github.com/mehah/otclient/blob/3c90b17379ad690f64d596c054aa75830a66fe9e/src/client/tile.cpp#L373
https://github.com/mehah/otclient/blob/3c90b17379ad690f64d596c054aa75830a66fe9e/src/client/tile.cpp#L881
https://github.com/mehah/otclient/blob/3c90b17379ad690f64d596c054aa75830a66fe9e/src/client/tile.h#L196

make sure they are the same and on the same line.

**Comment by oarsc** [comment]
I always do rebase all my changes on top of mehah/cache-for-all, there are some custom protocol changes that I have on my server that I can't avoid (as you can see in the video, I'm using a custom 7.73 version for it). However, I haven't touched a thing about drawing or illumination. The most critical part that could affect is a bigger viewport that I'm using..

Looks like, after logging, the server first sends the player outfit, and on a second message it switches to the Demon disguise...
So it hides the character because his player outfit is 1 dimensional.

On login player "Tester" outfit id is 131 (player outfit).
After a few milliseconds, it switches to 35 (demon outfit).
It's probably just a server thing, but I'll try to see more

However... see this:

https://user-images.githubusercontent.com/5507777/127640228-9ab92f96-88ef-4556-9db8-3f06950bcd2b.mp4

https://user-images.githubusercontent.com/5507777/127640890-897d58c2-ac70-4642-9b37-8422a00e3ced.mp4

https://user-images.githubusercontent.com/5507777/127641195-9908a0b0-913c-412b-8c7b-cfc3fa8f7204.mp4

On first logging, all visible characters are being drawn under roofs.
However, I think that even 1 dimensional character should be drawn so we could see his head..

Can you reproduce any of this?
I'm sorry I can't help more here, but don't have much knowledge about this part..

**Comment by mehah** [comment]
These problems you pointed out are easy to fix, I'm just worried about the previous one that I can't reproduce.

**Comment by oarsc** [comment]
Fix that and I'll look for some other things ;)
Maybe it also fixes something else

**Comment by mehah** [comment]
test now plz

**Comment by oarsc** [comment]
I couldn't find any problem with the creature sprite anymore, however there is still 2 things:
- Piles of crates/parcels are not being rendered either
- Green name when stopped moving while being hidden by a roof

https://user-images.githubusercontent.com/5507777/131168146-8088c12f-b4e3-4460-881a-94d8c4d9be98.mp4

**Comment by mehah** [comment]
I will be returning with the old code.

**Comment by mehah** [comment]
test now

**Comment by oarsc** [comment]
Sorry, I did more tests BEFORE your last commit (e2f7849) and I found something else.
It also happens AFTER e2f7849:
https://user-images.githubusercontent.com/5507777/131185453-5c1e5e89-aa71-43b8-ac59-7a55c4864967.mp4

I also reproduced this in ForgottenServer
https://user-images.githubusercontent.com/5507777/131185537-b29fc72c-5405-49a7-9e17-7b3ed3f65aab.mp4

I think the problem is when using a 1x1 sprite outfit. "New" ones are 2x2.

----------------------

After applying e2f7849, there is no problem with the "Green name" but this broke (this was working before):
https://user-images.githubusercontent.com/5507777/131186346-a7a1d1b6-5ddc-49c5-822a-a15e7f588c75.mp4
